src/Language/PureScript/CoreImp/Optimizer/MagicDo.hs

```diff
@@ -73,6 +73,8 @@ magicDo = inlineST . everywhere undo . everywhereTopDown convert
   -- Remove __do function applications which remain after desugaring
   undo :: AST -> AST
   undo (Return _ (App _ (Function _ (Just ident) [] body) [])) | ident == fnName = body
+  undo (App _ (App s1 (Function s2 Nothing [] (Block ss body)) []) []) =
```

Comment:
I'm not sure why this case is needed, but it makes the code harder to read in my opinion. `applyReturns` is only used in `convert` right now, and `undo` runs after the full traversal.

Can you give an example of why this is necessary? Is it possible to accomplish the same thing by modifying the `convert` function?